### PR TITLE
Fill command diagnostics

### PR DESCRIPTION
**Pull request by emmanueldenloye**
We've observed that while attempting to fill gaps (missing rows in each table) in the database, that gaps would still remain. 

We noticed upon further inspection that we were accessing the block header and payload endpoints incorrectly. Both can be accessed via chainweb-node's P2P API, however, it is only possible to get back at most 20 items and we had previously assumed we would always retrieve 360 items. We simply had to make sure to use chainweb-node's service API. 

We also noticed another small error in the way gaps were being calculated. Even with a full db, chainweb-data would proclaim that it had gaps of size 0 at the genesis block of chain.

### CODE REVIEW

File: haskell-src/lib/ChainwebData/Env.hs [hunk 1]
```diff
@@ -226,6 +226,7 @@ data BackfillArgs = BackfillArgs
 
 data FillArgs = FillArgs
   { _fillArgs_delayMicros :: Maybe Int
+  , _fillArgs_fillErrorLog :: Maybe FilePath
```
Comment (enobayram) [inline haskell-src/lib/ChainwebData/Env.hs:229]: Do we still need this argument now that we have fixed the issue? If we're planning to keep it, we need to have a good explanation of what this log file means. I.e. which kinds of errors are logged here.

File: haskell-src/lib/ChainwebData/Env.hs [hunk 2]
```diff
@@ -394,6 +395,9 @@ etlP = optional $ toETLEnv
 delayP :: Parser (Maybe Int)
 delayP = optional $ option auto (long "delay" <> metavar "DELAY_MICROS" <> help  "Number of microseconds to delay between queries to the node")
 
+logFile :: String -> Parser (Maybe FilePath)
+logFile message = optional $ strOption (long message <> metavar "LOG_FILE" <> help "File to log to")
```
Comment (enobayram) [inline haskell-src/lib/ChainwebData/Env.hs:399]:
```suggestion
fillErrorLogFile :: Parser (Maybe FilePath)
fillErrorLogFile = optional $ strOption (long "fill-error-log-file" <> metavar "PATH" <> help "File to write ??? kind of logs")
```
Is there any reason to receive the `long` name as an argument here?

File: haskell-src/exec/Chainweb/Worker.hs [hunk 3]
```diff
@@ -82,35 +83,71 @@ writes pool b ks ts es ss tf = P.withResource pool $ \c -> withTransaction c $ d
         --   (unDbHash $ _block_hash b)
         --   (map (const '.') ts)
 
-batchWrites :: P.Pool Connection -> [Block] -> [[T.Text]] -> [[Transaction]] -> [[Event]] -> [[Signer]] -> [[Transfer]] -> IO ()
-batchWrites pool bs kss tss ess sss tfs = P.withResource pool $ \c -> withTransaction c $ do
+batchWrites :: P.Pool Connection -> (ChainId, Int, Int) -> Maybe FilePath -> [Block] -> [[T.Text]] -> [[Transaction]] -> [[Event]] -> [[Signer]] -> [[Transfer]] -> IO ()
+batchWrites pool _chunkRange errorLogFile bs kss tss ess sss tfs = P.withResource pool $ \c -> withTransaction c $ do
+
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Worker.hs:88]:
```suggestion
batchWrites :: 
  P.Pool Connection -> 
  (ChainId, Int, Int) -> 
  Maybe FilePath -> 
  [Block] -> 
  [[T.Text]] -> 
  [[Transaction]] -> 
  [[Event]] -> 
  [[Signer]] -> 
  [[Transfer]] -> 
  IO ()
batchWrites pool _chunkRange errorLogFile bs kss tss ess sss tfs = 
  P.withResource pool $ \c -> withTransaction c $ do

```

File: haskell-src/exec/Chainweb/Server.hs [hunk 4]
```diff
@@ -198,7 +198,7 @@ scheduledUpdates env pool runFill fillDelay = forever $ do
 
     when runFill $ do
       logg Info "Filling missing blocks"
-      gaps env (FillArgs fillDelay)
+      gaps env (FillArgs fillDelay Nothing) -- TODO: This shouldn't be Nothing
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Server.hs:201]: Let's deal with this TODO one way or another

File: haskell-src/exec/Chainweb/Gaps.hs [hunk 5]
```diff
@@ -80,12 +87,108 @@ gapsCut env args cutBS = do
...
+    getNodeInfo manager serviceUrlScheme >>= \case
+      Left _err -> l Error "Error getting node info"
+      Right ni -> do
+        let pgc = PGInfo $ ConnectInfo
+              {
+                connectHost = "localhost"
+              , connectPort = 5432
+              , connectUser = "ubuntu"
+              , connectPassword = "somethingsecret"
+              , connectDatabase = "chainweb-data"
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Gaps.hs:165]: This doesn't look good

File: haskell-src/exec/Chainweb/Gaps.hs [hunk 6]
```diff
@@ -80,12 +87,108 @@ gapsCut env args cutBS = do
...
       let range = (ChainId (fromIntegral cid), l, h)
-      headersBetween env range >>= \case
-        Left e -> logger Error $ fromString $ printf "ApiError for range %s: %s" (show range) (show e)
+      let onCatch (e :: SomeException) = case errorLogFile of
+            Nothing -> do
+              logger Error $ fromString $ printf "Caught exception from headersBetween for range %s: %s" (show range) (show e)
+              pure $ Right []
+            Just fp -> withFileLogger fp Error $ do
+              S.logg Error $ fromString $ printf "Caught exception from headersBetween for range %s: %s" (show range) (show e)
+              pure $ Right []
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Gaps.hs:96]: I see a lot of code duplication of this kind, would be nice to deduplicate them (if we're planning to keep the errorLogFile)

File: haskell-src/exec/Chainweb/Worker.hs [hunk 7]
```diff
@@ -82,35 +83,71 @@ writes pool b ks ts es ss tf = P.withResource pool $ \c -> withTransaction c $ d
...
         $ insert (_cddb_blocks database) (insertValues bs)
         $ onConflict (conflictingFields primaryKey) onConflictDoNothing
       -- Write Pub Key many-to-many relationships if unique --
-      runInsert
-        $ insert (_cddb_minerkeys database) (insertValues $ concat $ zipWith (\b ks -> map (MinerKey (pk b)) ks) bs kss)
+
+      forM_ errorLogFile $ \fp -> withFileLogger fp Debug $ do
+        S.logg Debug $ "Over range: " <> fromString (show _chunkRange)
+        S.logg Debug $ fromString $ printf "Blocks returned: %s" $ show blocksReturned
+        S.logg Debug $ fromString $ printf "Blocks inserted: %d" $ length bs
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Worker.hs:99]:
```suggestion
      let 
        logInsertStats entityName inserted attempted = 
          forM_ errorLogFile $ \fp -> withFileLogger fp Debug $ do
            S.logg Debug $ "Over range: " <> fromString (show _chunkRange)
            S.logg Debug $ fromString $ printf "%s inserted: %s" entityName (show inserted)
            S.logg Debug $ fromString $ printf "%s attempted: %d" entityName attempted
      logInsertStats "Blocks" blocksReturned (length bs)
```

Seems like we can reuse this function for all the other cases

File: haskell-src/exec/Chainweb/Worker.hs [hunk 8]
```diff
@@ -164,9 +206,10 @@ writeBlocks env pool count bhs = do
               !sss = M.intersectionWith (\pl _ -> concat $ mkTransactionSigners . fst <$> _blockPayloadWithOutputs_transactionsWithOutputs pl) pls (makeBlockMap bhs')
               !kss = M.intersectionWith (\p _ -> bpwoMinerKeys p) pls (makeBlockMap bhs')
               err = printf "writeBlocks failed because we don't know how to work this version %s" version
+              chunkRange = (chain, minBH, maxBH)
           withEventsMinHeight version err $ \evMinHeight -> do
               let !tfs = M.intersectionWith (\pl bh -> mkTransferRows (fromIntegral $ _blockHeader_height bh) (_blockHeader_chainId bh) (DbHash $ hashB64U $ _blockHeader_hash bh) (posixSecondsToUTCTime $ _blockHeader_creationTime bh) pl evMinHeight) pls (makeBlockMap bhs')
-              batchWrites pool (M.elems bs) (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
+              batchWrites pool chunkRange (appendChain chain <$> errorLogFile) (M.elems bs) (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Worker.hs:212]:
```suggestion
              batchWrites pool chunkRange (appendChain chain <$> errorLogFile) (M.elems bs) 
                (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
```

**Review by enobayram — CHANGES_REQUESTED** [review]
Great job diagnosing and fixing this bug, which has been a source of inconvenience for `chainweb-data` operators.

File: haskell-src/exec/Chainweb/Gaps.hs (hunk 5 above)
Comment (emmanueldenloye) [reply haskell-src/exec/Chainweb/Gaps.hs]: I've taken this info out.

File: haskell-src/exec/Chainweb/Server.hs [hunk 9]
```diff
@@ -198,7 +198,7 @@ scheduledUpdates env pool runFill fillDelay = forever $ do
 
     when runFill $ do
       logg Info "Filling missing blocks"
-      gaps env (FillArgs fillDelay)
+      gaps env (FillArgs fillDelay) -- TODO: This shouldn't be Nothing
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Server.hs:201]:
```suggestion
      gaps env (FillArgs fillDelay)
```

File: haskell-src/exec/Chainweb/Worker.hs [hunk 10]
```diff
@@ -166,7 +181,8 @@ writeBlocks env pool count bhs = do
               err = printf "writeBlocks failed because we don't know how to work this version %s" version
           withEventsMinHeight version err $ \evMinHeight -> do
               let !tfs = M.intersectionWith (\pl bh -> mkTransferRows (fromIntegral $ _blockHeader_height bh) (_blockHeader_chainId bh) (DbHash $ hashB64U $ _blockHeader_hash bh) (posixSecondsToUTCTime $ _blockHeader_creationTime bh) pl evMinHeight) pls (makeBlockMap bhs')
-              batchWrites pool (M.elems bs) (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
+              batchWrites pool (M.elems bs)
+                (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Worker.hs:185]:
```suggestion
              batchWrites pool (M.elems bs) (M.elems kss) (M.elems tss) (M.elems ess) (M.elems sss) (M.elems tfs)
```
No need to introduce a whitespace diff

File: haskell-src/exec/Chainweb/Worker.hs [hunk 11]
```diff
@@ -149,8 +164,8 @@ writeBlocks env pool count bhs = do
       let ff bh = (hashToDbHash $ _blockHeader_payloadHash bh, _blockHeader_hash bh)
       retrying policy check (const $ payloadWithOutputsBatch env chain (M.fromList (ff <$> bhs')) id) >>= \case
         Left e -> do
-          logger Error $ fromString $ printf "Couldn't fetch payload batch for chain: %d" (unChainId chain)
-          logger Error $ fromString $ show e
+              logger Error $ fromString $ printf "Couldn't fetch payload batch for chain: %d" (unChainId chain)
+              logger Error $ fromString $ show e
```
Comment (enobayram) [inline haskell-src/exec/Chainweb/Worker.hs:168]:
```suggestion
          logger Error $ fromString $ printf "Couldn't fetch payload batch for chain: %d" (unChainId chain)
          logger Error $ fromString $ show e
```

**Review by enobayram — APPROVED** [review]
LGTM! Great work with the fixes!

**Pull request merged**